File: java-algorithm/src/main/java/com/leetcode/Problem606.java
package com.leetcode;

import com.common.TreeNode;

/**
 * 606. Construct String from Binary Tree
 * https://leetcode.com/problems/construct-string-from-binary-tree/
 */
class Problem606 {
    public String tree2str(TreeNode root) {
        if (root == null) return "";
        String left = tree2str(root.left);
        String right = tree2str(root.right);
        if (root.left != null && root.right != null)
            return String.format("%d(%s)(%s)", root.val, left, right);
        if (root.left != null)
            return String.format("%d(%s)", root.val, left);
        if (root.right != null)
            return String.format("%d()(%s)", root.val, right);
        return "" + root.val;
    }
}

File: python-algo/leetcode/problem_606.py
"""606. Construct String from Binary Tree
https://leetcode.com/problems/construct-string-from-binary-tree/
"""
from typing import Optional

from common.tree_node import TreeNode


class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        def dfs(node: TreeNode) -> str:
            if not node:
                return ""
            ret = str(node.val)
            if not node.left and not node.right:
                return ret
            if not node.left and node.right:
                return ret + '()(' + dfs(node.right) + ')'
            if node.left and not node.right:
                return ret + '(' + dfs(node.left) +')'
            else:
                return ret + '(' + dfs(node.left) +')' + '(' + dfs(node.right) + ')'
        return dfs(root)
